### ISSUE TITLE
memory-packing pass invalidates Haskell code

### ISSUE THREAD

**Issue by TerrorJack** [WebAssembly/binaryen#3190]
[Here](https://drive.google.com/file/d/1499FaNWW8GYGs5BM-fiodXC1ePcBmkhB/view?usp=sharing) is a tarball of the JavaScript/WebAssembly code. It can be run with `node unicode.mjs`, which will output `True` to the console. (Tested with `node-v14.13.0`)

If we run `wasm-opt -O0 --memory-packing unicode.wasm -o unicode.wasm`, the `memory-packing` pass is run which invalidates the wasm code, making it OOM and crash without warning. This is tested using `binaryen` `version_97`. We also confirmed that other passes combined don't cause the crash (the ones when optimization level is 4 and shrinking level is 2).

**Comment by tlively** [comment]
Can you elaborate about how you are running the code and observing OOM and crashing? Is this in Node or in a browser? Does the issue reproduce if you do `wasm-opt --memory-packing unicode.wasm -o unicode.wasm` (without `-O0`)?

**Comment by TerrorJack** [comment]
> Can you elaborate about how you are running the code and observing OOM and crashing? Is this in Node or in a browser?

This is in node. The code is run with `node ./unicode.mjs`, which will compile and instantiate the `unicode.wasm` file and run it.

> Does the issue reproduce if you do wasm-opt --memory-packing unicode.wasm -o unicode.wasm (without -O0)?

It doesn't, since without the `-O0` flag `wasm-opt` doesn't do any work.

```
terrorjack@build01:~/asterius/test/unicode$ wasm-opt --debug --memory-packing unicode.wasm -o unicode.wasm
warning: no passes specified, not doing any work
```

**Comment by tlively** [comment]
Huh, the `--debug` flag makes it think that `--memory-packing` isn't a pass for some reason. Very strange, but also a separate issue. I am able to reproduce the issue locally with `wasm-opt --memory-packing unicode.wasm -o unicode.wasm`, so I will look into it further.

**Comment by kripken** [comment]
Maybe `wasm-reduce` can help here? If you try it and it doesn't help get a nice small testcase, let me know, I don't think it's been tested much on memory and data segment stuff (since most bugs are in code...).

**Comment by tlively** [comment]
Well the reducer is chugging along removing one byte at a time from the data segments 😆 I'll let it run for a few days at the most and see where we get.

**Comment by kripken** [comment]
Ah, we should probably add exponential growth & backoff to that part of the reducer, like we have for other things (exports, functions).

**Comment by TerrorJack** [comment]
Thanks for reminding about `wasm-reduce` :+1: 
I've managed to get it running on our build machine and let's see how it goes. I had to use `-f` to skip the canonical check (no idea why it fails), and it seems that mutating the wasm module specified by `-t` isn't supported by `wasm-reduce` (so the test script needs to copy that to elsewhere for `wasm-opt`).

**Comment by TerrorJack** [comment]
@tlively Maybe it's a bit late already, but still I discovered a trick to speed up iterations so maybe it can help on your end: supply `--max-heap-size=6` to `node`, so after `wasm-opt` reduction it should crash almost instantly instead of having to wait for a long time.

**Comment by tlively** [comment]
Nice, thanks for the hint! My wasm-reduce is still going, so I think I will have to stop it and make some improvements to get it to finish in a reasonable time. By the way, if you had to use `-f` for wasm-reduce, it's probably not preserving the property you care about. Here's my invocation:

```
../bin/wasm-reduce unicode.wasm -t t.wasm -w w.wasm -cmd="cp t.wasm t2.wasm && timeout 2 node unicode.mjs &> /dev/null && echo 'success' && ../bin/wasm-opt --memory-packing t.wasm -o t2.wasm && timeout 2 node unicode.mjs &> /dev/null && echo 'fail'" --timeout=4
```

I edited unicode.wasm.mjs to load from t2.wasm. When the property is preserved, the reduce command will print "success" but not "fail".

**Comment by TerrorJack** [comment]
> By the way, if you had to use -f for wasm-reduce, it's probably not preserving the property you care about.

`wasm-reduce` has a check which runs `wasm-opt` with default flags and sees if the runner command behaves the same, according to [source](https://github.com/WebAssembly/binaryen/blob/version_97/src/tools/wasm-reduce.cpp#L1193). And unfortunately in this case, default level should already have included `memory-packing` pass which of course triggers the bug and changes the behavior, thus the `-f`.

**Comment by TerrorJack** [comment]
[Here](https://drive.google.com/file/d/193gYPw7v282eiCSLL6jou14MgiuiVGbb/view?usp=sharing) is the minified wasm module after a week's `wasm-reduce` run.

**Comment by tlively** [comment]
@TerrorJack This has been fixed both in Asterius and in #3239, right? Can this issue be closed?

**Comment by TerrorJack** [comment]
@tlively Yes indeed, thanks for reminding! I'm closing this one.
